### ISSUE TITLE
NullPointerException on empty result set

### ISSUE THREAD

**Issue by moxious** [neo4j-contrib/neo4j-spark-connector#227]
Related to the fix of this issue: https://github.com/neo4j-contrib/neo4j-spark-connector/issues/200 I think we have a regression.   Please note about this issue that what I'm connecting to is a 3.5 causal cluster, so I suspect this is 3.5 specific.  The same query run against the built-in standalone 4.1.2 instance in the notebooks repo does *not* reproduce this error, and works as expected.

My code: 

```
%spark

spark.read.format("org.neo4j.spark.DataSource").option("url", "neo4j://(REMOVED):7687")
    .option("authentication.type", "basic")
    .option("authentication.basic.username", "neo4j")
    .option("authentication.basic.password", "(REMOVED)")
    .option("query", "MATCH (i:DO_NOT_EXIST) RETURN i").load().show()
```

This results in:

```
java.lang.NullPointerException
  at org.neo4j.spark.service.SchemaService.getReturnedColumns(SchemaService.scala:226)
  at org.neo4j.spark.service.SchemaService.structForQuery(SchemaService.scala:194)
  at org.neo4j.spark.service.SchemaService.struct(SchemaService.scala:239)
  at org.neo4j.spark.reader.Neo4jDataSourceReader$$anonfun$2.apply(Neo4jDataSourceReader.scala:25)
  at org.neo4j.spark.reader.Neo4jDataSourceReader$$anonfun$2.apply(Neo4jDataSourceReader.scala:24)
  at org.neo4j.spark.reader.Neo4jDataSourceReader.callSchemaService(Neo4jDataSourceReader.scala:34)
  at org.neo4j.spark.reader.Neo4jDataSourceReader.<init>(Neo4jDataSourceReader.scala:24)
  at org.neo4j.spark.DataSource.createReader(DataSource.scala:17)
  at org.neo4j.spark.DataSource.createReader(DataSource.scala:13)
  at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$SourceHelpers.createReader(DataSourceV2Relation.scala:155)
  at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation$.create(DataSourceV2Relation.scala:172)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:204)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:167)
  ... 45 elided
```

**Comment by moxious** [comment]
@utnaf replicate with this branch: https://github.com/utnaf/spark-connector-notebooks/tree/neo4j-3.5 .  I won't PR a 3.5 version of the notebooks, but if you start this remember to remove your neo4j 4 data files before launching

Replication does not require a cluster

**Comment by utnaf** [comment]
We'll have a slight improve of the error message when merging #229 

```
java.lang.IllegalArgumentException: Cannot compute the StructType for the provided query type,
please check the params or the query
```

problem here is that in Neo4j 3.5 we can't get an actual list of returned data from the query plan object in the driver (see discussion with Michael Simons). 

We can get a list of whole the identifier in the query, for example, the query

```
MATCH (i:Item) RETURN i.name as name, id(i) as id
```
will have as identifiers: `i, name, id`. 

We are a bit stuck here I guess, possible "solutions":
- be more specific with the error (we can't compute the schema if the dataset is empty on neo4j 3.5)
- add the famous option to specify the schema `.option("query.schema", "name:String,id:Long"), all comes down to this
- adding a warning that tells the user that the query result is empty so stuff might not work
- suggestions?

wdyt? @moxious @conker84

**Comment by moxious** [comment]
If you can do the first solution with a very clear error message, that's probably good enough.  An error like, "Unable to compute the resulting schema; this may mean your result set is empty or your version of Neo4j does not permit schema inference for empty sets".   This would at least explain the issue, and allow us to suggest alternatives to the person

**Comment by utnaf** [comment]
Fixed in #229 by this commit https://github.com/neo4j-contrib/neo4j-spark-connector/pull/229/commits/f0e9b71e69983cbf5532f34be67ce3fb510d8a92

Fun fact: `MATCH (n) RETURN n` doesn't compute the schema, `MATCH (n) RETURN id(n)` does compute the schema.

This remember me of this bit of docs

> We recommend that individual property fields be returned, rather than returning graph entity (node, relationship, and path) types. This best maps to Spark’s type system and yields best results. So instead writing this MATCH (p:Person) RETURN p please write this: MATCH (p:Person) RETURN id(p) as id, p.name as name. If your query returns a graph entity please use the labels or relationship modes instead.

You can found it here http://5.9.211.195:8000/neo4j-spark-docs/1.0.0/reading.html#bookmark-read-query

We are covered anyways with a clean error.
